### PR TITLE
DPO Loss refactor

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation
- [x] refactor

I was initially looking at this refactor to add SimPO into the DPO recipe, but got slightly carried away. I think we might need a separate SimPO recipe down the line anyway, since DPO-style losses use a reference model, whereas SimPO doesn't.

This PR refactors the DPO loss module into separate classes for each of the loss types it supported. Each separate loss is now documented with a reference to its corresponding paper, a little intuition about how it works, and comes with a corresponding unit test.

#### Test plan

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add unit tests for any new functionality
- [x] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

### CODE REVIEW

**Review by RdoubleA — APPROVED** [review]
Excellent refactor. I honestly don't have any concerns here.

As a DPO noob, do you mind adding some details in the docstrings about the training setup for each of these losses? Like what is the policy/reference model, and the core value prop of DPO compared to PPO is that it does not require a reward model. 

Also, are we completely removing the kto pair loss?

**Comment by SalmanMohammadi** [comment]
> Also, are we completely removing the kto pair loss?

Yeah, sorry, I forgot to point this out. The reference implementation in TRL seems to differ enough (it uses a separate trainer with a different dataset format), and I couldn't see tests for the KTO loss in the original DPO PR. I personally would re-consider this when we have tests against a reference implementation.
